File: src-tauri/src/indicator.rs
//! System-wide "recording indicator" overlay window.
//!
//! A second, frameless, always-on-top [`WebviewWindow`] (label
//! [`INDICATOR_LABEL`]) — a small bottom-center pill that shows ONLY a recording
//! cue while capture is live: an audio-reactive waveform (WhisperFlow-style — no
//! record dot, no timer). Deliberately NO transcription text. It's for users who
//! want a clear "you're recording" indicator WITHOUT the live-caption overlay,
//! and it works even when live preview is fully off.
//!
//! This is a faithful, independent MIRROR of the live-preview overlay (see
//! [`crate::overlay`]): same machinery (frameless, decorations off, always-on-top,
//! skip-taskbar, OPAQUE — a transparent always-on-top WebView2 window hard-crashes
//! on Windows, so it's an opaque themed panel), reconciled against its own config
//! flag `interface.recording_indicator`. It is entirely separate from the caption
//! overlay: a different window label, a different config flag, and its own
//! per-label `tauri-plugin-window-state` geometry — either, both, or neither can
//! run at once.
//!
//! ## How recording state reaches it
//! Nothing extra is needed: [`crate::events`] re-emits every daemon event with
//! `app.emit("daemon-event", …)`, and Tauri's `Emitter::emit` broadcasts to
//! **all** webviews. So the moment this window exists it receives the same
//! `recording_started` / `audio_level_sample` / `recording_stopped` stream the
//! main window does, and `indicator.ts` drives show/hide + the waveform from it.
//! No transcription/preview is involved.
//!
//! ## Window lifecycle
//! * Created **hidden** at startup when the setting is on (so the very first
//!   recording can show it instantly without a cold window build), and on a
//!   config save that flips the setting on.
//! * Destroyed on a config save that flips it off.
//! * `indicator.ts` shows it on `recording_started` and hides it on
//!   `recording_stopped`/`cancelled`/`deleted` — Rust never forces visibility.
//!
//! ## Position persistence
//! The window is draggable (`indicator.ts` repositions it manually with
//! `setPosition` on pointer drag — NOT a `data-tauri-drag-region`, whose OS modal
//! move-loop can freeze an always-on-top window's shared event loop). Its position
//! is remembered across runs by `tauri-plugin-window-state`, which saves/restores
//! geometry per window label automatically. On first ever creation (no saved
//! state) we place it bottom-center of the primary monitor.

use tauri::{AppHandle, Manager, WebviewWindowBuilder};

/// The label of the recording-indicator [`WebviewWindow`]. Must match the label
/// used in `frontend/src/indicator.ts` (`getCurrentWindow()` there resolves to
/// this) and the `windows` allowlist in `src-tauri/capabilities/default.json`.
pub const INDICATOR_LABEL: &str = "recording-indicator";

/// Indicator width (logical px). Sized snugly to the centered audio waveform
/// (13 bars × 3px + 3px gaps ≈ 75px) plus the card's 16px side padding — so the
/// pill hugs the bars instead of leaving dead space around them. Fixed (the
/// height is pinned too): a tiny status cue, not a resizable panel. Kept in sync
/// with `indicator.css` (`.ri-card` padding + `.ri-wave` bar count/width).
const INDICATOR_W: f64 = 112.0;
/// Indicator height (logical px). Snug around the 22px waveform with a little
/// vertical breathing room inside the rounded pill.
const INDICATOR_H: f64 = 32.0;
/// Inset from the bottom of the work area for the first-run placement.
const BOTTOM_MARGIN: f64 = 96.0;
/// Off-screen sentinel start position (mirrors [`crate::overlay`]): a never-saved
/// window starts off every monitor so first-run placement is detectable as "not
/// on any monitor", instead of guessing from coordinate signs (which mis-flagged
/// monitors arranged left of / above the primary).
const OFFSCREEN_SENTINEL: f64 = -32000.0;

/// Whether the indicator window currently exists.
pub fn exists(app: &AppHandle) -> bool {
    app.get_webview_window(INDICATOR_LABEL).is_some()
}

/// Create the indicator window (hidden) if it doesn't already exist. Idempotent.
///
/// The window is frameless, always-on-top, skips the taskbar, and is a fixed
/// small size (NOT resizable — it's a tiny pill). It is OPAQUE, never transparent
/// (a transparent + always-on-top + frameless WebView2 window hard-crashes the
/// whole app on some Windows/WebView2 builds — see [`crate::overlay`]). It starts
/// **hidden**; `indicator.ts` reveals it when a recording starts. Returns early
/// (and logs) on failure so a broken indicator never blocks the app.
pub fn ensure(app: &AppHandle) {
    if exists(app) {
        return;
    }

    // `indicator.html` lives at the frontend root, so the in-app URL is just
    // `indicator.html` (Vite emits it as a sibling of `index.html` — see the
    // multi-input `rollupOptions` in vite.config.ts). The same relative path
    // works for both the dev server and the bundled `frontendDist`.
    let url = tauri::WebviewUrl::App("indicator.html".into());

    let builder = WebviewWindowBuilder::new(app, INDICATOR_LABEL, url)
        .title("Phoneme Recording Indicator")
        .inner_size(INDICATOR_W, INDICATOR_H)
        // Start off every monitor (hidden) so first-run placement is detectable
        // as "not on any monitor"; the window-state plugin overrides this with a
        // remembered position when one exists. See `place_default_if_unpositioned`.
        .position(OFFSCREEN_SENTINEL, OFFSCREEN_SENTINEL)
        // Fixed small pill: pin BOTH axes by making min == max == the inner size,
        // so the user can't resize it into something odd. Position is still
        // remembered by tauri-plugin-window-state.
        .min_inner_size(INDICATOR_W, INDICATOR_H)
        .max_inner_size(INDICATOR_W, INDICATOR_H)
        .resizable(false)
        .decorations(false)
        // NOT transparent: a transparent + always-on-top + frameless WebView2
        // window hard-crashes the whole app on some Windows/WebView2 builds when
        // shown. The indicator is an opaque themed pill instead — see
        // indicator.css.
        .always_on_top(true)
        .skip_taskbar(true)
        // Don't steal focus when it pops up mid-recording.
        .focused(false)
        .visible(false);

    let window = match builder.build() {
        Ok(w) => w,
        Err(e) => {
            tracing::warn!(error = %e, "failed to create recording-indicator window");
            return;
        }
    };

    // Force the correct fixed size even if `tauri-plugin-window-state` restored a
    // stale geometry from an older build (the programmatic restore bypasses the
    // min/max above). Without this, a pre-existing 150×36 saved size would leave
    // dead space around the snug 112×32 pill. Best-effort.
    let _ = window.set_size(tauri::LogicalSize::new(INDICATOR_W, INDICATOR_H));

    // First-run placement: the window-state plugin overwrites the off-screen
    // sentinel only when it restored a saved position, so "is it on a connected
    // monitor?" tells us whether to nudge to bottom-center.
    place_default_if_unpositioned(&window);

    tracing::info!("recording-indicator window created (hidden)");
}

/// Destroy the indicator window if it exists. Idempotent. Called when the setting
/// is turned off so we don't keep an invisible window (and its webview) around.
pub fn destroy(app: &AppHandle) {
    if let Some(w) = app.get_webview_window(INDICATOR_LABEL) {
        if let Err(e) = w.close() {
            tracing::warn!(error = %e, "failed to close recording-indicator window");
        }
    }
}

/// Reconcile the indicator window against the desired enabled state: create it
/// (hidden) when on, destroy it when off. Safe to call repeatedly. Used at
/// startup and after every config save/profile switch.
pub fn sync(app: &AppHandle, enabled: bool) {
    if enabled {
        ensure(app);
    } else {
        destroy(app);
    }
}

/// Place the indicator at the bottom-center of its monitor's work area, but only
/// when there's no remembered position to honor — detected by the window not
/// sitting on any currently-connected monitor (the off-screen builder sentinel on
/// first run, or a saved position whose monitor was unplugged). A restored
/// position on ANY monitor (including ones arranged left of / above the primary,
/// at negative coordinates) is respected. Best-effort.
fn place_default_if_unpositioned(window: &tauri::WebviewWindow) {
    if position_is_on_a_monitor(window) {
        return; // a remembered, on-screen position was restored — respect it
    }

    let Ok(Some(monitor)) = window.current_monitor() else {
        return;
    };
    let scale = monitor.scale_factor();
    let size = monitor.size().to_logical::<f64>(scale);
    let pos = monitor.position().to_logical::<f64>(scale);

    let x = pos.x + (size.width - INDICATOR_W) / 2.0;
    let y = pos.y + size.height - INDICATOR_H - BOTTOM_MARGIN;
    let _ = window.set_position(tauri::LogicalPosition::new(x.max(pos.x), y.max(pos.y)));
}

/// Whether the window's top-left corner falls within any connected monitor's
/// bounds. Distinguishes a restored on-screen position from the off-screen
/// first-run sentinel (or a now-disconnected monitor). On any error reading the
/// position or enumerating monitors, returns `false` so the caller places the
/// window rather than leaving it off-screen.
fn position_is_on_a_monitor(window: &tauri::WebviewWindow) -> bool {
    let Ok(p) = window.outer_position() else {
        return false;
    };
    let Ok(monitors) = window.available_monitors() else {
        return false;
    };
    monitors.iter().any(|m| {
        let mp = m.position();
        let ms = m.size();
        let right = mp.x + ms.width as i32;
        let bottom = mp.y + ms.height as i32;
        p.x >= mp.x && p.x < right && p.y >= mp.y && p.y < bottom
    })
}

File: frontend/src/indicator.ts
// Entry point for the system-wide "recording indicator" overlay window.
//
// This runs in its OWN Tauri WebviewWindow (label "recording-indicator"), created
// at runtime by the tray when the "recording indicator" setting is on (see
// src-tauri/src/indicator.rs). It is a SLIM sibling of the live-preview overlay
// (src/overlay.ts): a tiny always-on-top pill showing ONLY an audio-reactive
// waveform — no record dot, no timer, no transcription text (WhisperFlow-style).
// It needs no live preview at all and works even when live preview is fully off.
//
// It listens to the same global `daemon-event` stream the main window does:
// shows on `recording_started`, drives the waveform from `audio_level_sample`,
// and hides on `recording_stopped`/`cancelled`/`deleted` (once no track remains,
// mirroring overlay.ts's meeting-track bookkeeping). Fully independent of the
// caption overlay — a different window, flag, and saved geometry; either, both,
// or neither can run.
import "./styles/theme.css";
import "./styles/indicator.css";
import { listen } from "@tauri-apps/api/event";
import { getCurrentWindow } from "@tauri-apps/api/window";
import { LogicalPosition } from "@tauri-apps/api/dpi";
import { invoke } from "@tauri-apps/api/core";

const root = document.getElementById("indicator-root")!;
// Just the waveform, centered in a small fixed pill (size pinned in indicator.rs).
root.innerHTML = `
  <div class="ri-card">
    <span class="ri-wave" id="ri-wave" aria-hidden="true"></span>
  </div>
`;

const win = getCurrentWindow();

// ── "It hears me" waveform ───────────────────────────────────────────────────
// A row of bars driven by the daemon's audio_level_sample events (cheap mic RMS,
// no transcription). Built once; heights animate via CSS transform. This is the
// same perceptual sqrt-curve + gain the caption overlay uses (src/overlay.ts).
const WAVE_BARS = 13;
const waveEl = document.getElementById("ri-wave") as HTMLElement;
for (let i = 0; i < WAVE_BARS; i++) {
  const b = document.createElement("span");
  b.className = "ri-wave-bar";
  waveEl.appendChild(b);
}
const waveBars = Array.from(waveEl.querySelectorAll<HTMLElement>(".ri-wave-bar"));
const waveRing: number[] = new Array(WAVE_BARS).fill(0);

function pushLevel(level: number) {
  const raw = Math.max(0, Math.min(1, Number.isFinite(level) ? level : 0));
  // Speech RMS sits low (~0.05–0.3), so a linear bar barely twitches. A perceptual
  // sqrt curve + a little gain makes normal speech visibly drive the bars, while
  // the clamp still caps loud peaks at full height.
  const v = Math.min(1, Math.sqrt(raw) * 1.2);
  waveRing.push(v);
  waveRing.shift();
  for (let i = 0; i < waveBars.length; i++) {
    // 0.15 floor so the bars are always visible while active.
    waveBars[i].style.transform = `scaleY(${(0.15 + waveRing[i] * 0.85).toFixed(3)})`;
  }
}
function resetWave() {
  waveRing.fill(0);
  waveBars.forEach((b) => (b.style.transform = "scaleY(0.15)"));
}
resetWave();

// Apply the saved theme so the pill matches the app's look. Falls back to the CSS
// default if the config read fails — the indicator must never block on it.
void (async () => {
  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const cfg = await invoke<any>("read_config");
    if (cfg?.interface?.theme) {
      document.documentElement.setAttribute("data-theme", cfg.interface.theme);
    }
  } catch {
    /* keep CSS defaults */
  }
})();

// ── Window visibility lifecycle ──────────────────────────────────────────────
async function showIndicator() {
  document.body.classList.add("ri-live");
  try {
    await win.show();
    // Re-assert always-on-top each show: an app going fullscreen can steal the
    // top spot, and re-showing is the natural moment to reclaim it.
    await win.setAlwaysOnTop(true);
  } catch {
    /* window may be mid-teardown */
  }
}

function hideIndicator() {
  // Crisp single-step hide — no dim stage, no fade (matches the caption overlay's
  // teardown). Settle the waveform for the next show.
  document.body.classList.remove("ri-live");
  resetWave();
  void win.hide().catch(() => {});
}

// ── Meeting-track bookkeeping ────────────────────────────────────────────────
// A meeting has TWO tracks (mic + system) that each emit recording_started /
// stopped. Mirror overlay.ts: only hide once NO track remains, so a stop on one
// track doesn't tear the indicator down while the other is still live. Single
// recordings register no tracks, so they hide on their own stop as before.
const meetingTracks = new Map<string, string>();

// ── Daemon event stream ──────────────────────────────────────────────────────
// eslint-disable-next-line @typescript-eslint/no-explicit-any
void listen<any>("daemon-event", async (e) => {
  const p = e.payload;
  switch (p?.event) {
    case "recording_started": {
      // Off-switch authoritative at show time: never auto-show when
      // `interface.recording_indicator` is off, even if this window still exists
      // (created earlier, then disabled before its destroy landed).
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      let cfg: any = null;
      try { cfg = await invoke<any>("read_config"); } catch { /* assume enabled */ }
      if (cfg && !cfg.interface?.recording_indicator) break;
      if (p.meeting_id && typeof p.track === "string") {
        meetingTracks.set(p.id, p.track);
      } else {
        meetingTracks.clear();
      }
      resetWave();
      await showIndicator();
      break;
    }
    case "audio_level_sample":
      // Drive the "it hears me" waveform bars. Cheap; runs for any capture.
      pushLevel(typeof p.level === "number" ? p.level : 0);
      break;
    case "recording_stopped":
    case "recording_cancelled":
    case "recording_deleted":
      meetingTracks.delete(p.id);
      // Only hide once every track has stopped (single recordings have no tracks
      // registered, so they hide immediately); a fresh recording_started re-shows.
      if (meetingTracks.size === 0) {
        hideIndicator();
      }
      break;
  }
});

// ── Position persistence ─────────────────────────────────────────────────────
// tauri-plugin-window-state restores geometry per window label, but it only SAVES
// on a graceful app exit — a crash, force-kill, or dev rebuild loses any drag
// since launch. Save explicitly (debounced) after each move so the indicator's
// placement survives anything.
let saveTimer: number | null = null;
function queueStateSave() {
  if (saveTimer !== null) clearTimeout(saveTimer);
  saveTimer = window.setTimeout(() => {
    saveTimer = null;
    void invoke("save_window_state").catch(() => {
      /* non-fatal — exit-time save still applies */
    });
  }, 600);
}
void win.onMoved(() => queueStateSave());

// ── Manual window dragging ───────────────────────────────────────────────────
// Drag manually rather than via data-tauri-drag-region: the OS modal move-loop
// that drag-region triggers can freeze the shared Tauri event loop on an
// always-on-top frameless window (the "preview hangs the app when I move it"
// crash — see overlay.ts). Instead track the pointer and reposition with
// setPosition, coalesced to one per animation frame.
const card = root.querySelector<HTMLElement>(".ri-card")!;
let dragging = false;
let originX = 0; // window's logical-x at grab time
let originY = 0; // window's logical-y at grab time
let grabX = 0; // pointer screen-x at grab time (logical/CSS px)
let grabY = 0; // pointer screen-y at grab time (logical/CSS px)
let nextX = 0;
let nextY = 0;
let rafPending = false;

function flushDrag() {
  rafPending = false;
  if (!dragging) return;
  void win.setPosition(new LogicalPosition(nextX, nextY)).catch(() => {});
}

card.addEventListener("pointerdown", async (e) => {
  if (e.button !== 0) return; // left button only
  // Capture the grab point synchronously, before any await, so the reference is
  // the true press location even if the position read below is slow.
  grabX = e.screenX;
  grabY = e.screenY;
  e.preventDefault();
  try {
    const scale = await win.scaleFactor();
    const pos = await win.outerPosition(); // physical px → logical
    originX = pos.x / scale;
    originY = pos.y / scale;
  } catch {
    return; // window mid-teardown — leave it be
  }
  dragging = true;
  try {
    card.setPointerCapture(e.pointerId);
  } catch {
    /* capture is best-effort */
  }
});

card.addEventListener("pointermove", (e) => {
  if (!dragging) return;
  // screenX/Y are logical (CSS) px, matching LogicalPosition — no DPR math.
  nextX = originX + (e.screenX - grabX);
  nextY = originY + (e.screenY - grabY);
  if (!rafPending) {
    rafPending = true;
    requestAnimationFrame(flushDrag);
  }
});

function endDrag(e: PointerEvent) {
  if (!dragging) return;
  dragging = false;
  try {
    card.releasePointerCapture(e.pointerId);
  } catch {
    /* may already be released */
  }
  queueStateSave(); // persist the final resting position
}
card.addEventListener("pointerup", endDrag);
card.addEventListener("pointercancel", endDrag);
